### PR TITLE
bears/pug: Add PugLintBear

### PR DESCRIPTION
**Pull request by yash-nisar**
Closes https://github.com/coala/coala-bears/issues/290

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by yash-nisar** [comment]
CC @Makman2

File: bears/pug/PugLintBear.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            The location of the ``.pug-lintrc`` config file.
+        """
+        return ('--config',
+                puglint_config if puglint_config else config_file,
+                '--reporter', 'inline', filename)
+
+    def process_output(self, output, filename, file):
+        stdout, stderr = output
+        yield from self.process_output_regex(stderr, filename, file,
+                                             self._output_regex)
```
Comment (Makman2) [inline bears/pug/PugLintBear.py]: why you don't use `output_format=regex`?

File: tests/pug/puglint_test_files/multiple_line_breaks.pug [hunk 2]
```diff
@@ -0,0 +1,4 @@
+div
+
+
+div
```
Comment (Makman2) [inline tests/pug/puglint_test_files/multiple_line_breaks.pug]: are the missing \n at EOF on purpose?

File: bears/pug/PugLintBear.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+from dependency_management.requirements.NpmRequirement import NpmRequirement
+
+
+@linter(executable='pug-lint',
+        use_stdout=True,
+        use_stderr=True)
+class PugLintBear:
+    """
+    A configurable linter and style checker for Pug (formerly Jade) that is a
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: """``Pug``"'' (formerly "``jade``"

Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: enclose them within backsticks

File: bears/pug/PugLintBear.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                        require_id_literals_before_attributes: bool=False,
+                        require_lower_case_attributes: bool=True,
+                        require_lower_case_tags: bool=True,
+                        require_spaces_inside_attribute_brackets: bool=False,
+                        require_strict_equality_operators: bool=True,
+                        validate_div_tags: bool=True,
+                        validate_extensions: bool=True,
+                        validate_self_closing_tags: bool=True,
+                        preferred_quotation: str='\'',
+                        max_lines_per_file: int=None,
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: change ``max_lines_per_file`` -> ``max_file_length``, for consistency.

File: bears/pug/PugLintBear.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            static values.
+            For example: If set to ``True``, prefer ``span.foo`` over
+            ``span(class='foo')``.
+        :param prohibit_class_literals_before_attributes:
+            When ``True``, prefer all attribute blocks to be written before
+            any class literals.
+            For example: If set to ``True``, prefer
+            ``input(type='text').class`` over ``input.class(type='text')``.
+        :param prohibit_class_literals_before_id_literals:
+            When ``True``, prefer all ID literals to be written before any
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: ^^

File: bears/pug/PugLintBear.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+            When ``True``, all class literals should be written before any
+            ID literals.
+        :param require_id_literals_before_attributes:
+            When ``True``, all ID literals must be written before any
+            attribute blocks.
+        :param require_lower_case_attributes:
+            When ``True``, all attributes should be written in lower case.
+            For example: If set to ``True``, prefer ``div(class='class')``
+            over ``div(Class='class')``.
+        :param require_lower_case_tags:
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: enforce_lower_case_tags

File: bears/pug/PugLintBear.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+            When ``True``, all tags must be written in lower case.
+            For example: If set to ``True``, prefer ``div(class='class')``
+            over ``Div(class='class')``.
+        :param require_spaces_inside_attribute_brackets:
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: enforce_spaces_inside_attribute_brackets

File: bears/pug/PugLintBear.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            When ``True``, enforce space after opening attribute bracket and
+            before closing attribute bracket.
+        :param require_strict_equality_operators:
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: enforce_...

File: bears/pug/PugLintBear.py (hunk 1 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Done :tada:

File: tests/pug/puglint_test_files/multiple_line_breaks.pug (hunk 2 above)
Comment (yash-nisar) [inline tests/pug/puglint_test_files/multiple_line_breaks.pug]: Nope, my bad.

File: bears/pug/PugLintBear.py (hunk 3 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Done :+1:

File: bears/pug/PugLintBear.py (hunk 4 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: We have named the setting as `max_lines_per_file` in `LineCountBear`, so I think we should keep `max_lines_per_file` for consistency of the settings throughout.

File: bears/pug/PugLintBear.py (hunk 6 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Done :+1:

File: bears/pug/PugLintBear.py (hunk 7 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Done :+1:

File: bears/pug/PugLintBear.py (hunk 8 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Done :+1:

File: tests/pug/puglint_test_files/multiple_line_breaks.pug (hunk 2 above)
Comment (yash-nisar) [inline tests/pug/puglint_test_files/multiple_line_breaks.pug]: Fixed. :+1:

**Comment by yash-nisar** [comment]
Updated @Makman2

File: bears/pug/PugLintBear.py (hunk 4 above)
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: there is the ``TailorBear `` that has ``max_file_length``

Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Should we open an issue for that ?

Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: @Makman2 , @yash-nisar let's vote :smile: and get rid of the other setting 
i am for using ``max_file_length`` as we have ``max_line_length``

Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Nvm, doing so right away :smile:

Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Haha okay, lets vote :+1:

Comment (Makman2) [inline bears/pug/PugLintBear.py]: `max_file_length` might refer to the number of chars inside a document, `max_lines_per_file` is less ambigious^^ So imo `max_lines_per_file` is better 👍

File: bears/pug/PugLintBear.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+                        enforce_class_literals_before_id_literals: bool=False,
+                        enforce_id_literals_before_attributes: bool=False,
+                        enforce_lower_case_attributes: bool=True,
+                        enforce_lower_case_tags: bool=True,
+                        require_spaces_inside_attribute_brackets: bool=False,
+                        require_strict_equality_operators: bool=True,
+                        validate_div_tags: bool=True,
+                        validate_extensions: bool=True,
+                        validate_self_closing_tags: bool=True,
+                        preferred_quotation: str='\'',
```
Comment (Makman2) [inline bears/pug/PugLintBear.py]: use `"'"` instead of `'\''`

File: bears/pug/PugLintBear.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+        :param prohibit_class_attribute_with_static_value:
+            When ``True``, prefer class literals over class attributes with
+            static values.
+            For example: If set to ``True``, prefer ``span.foo`` over
+            ``span(class='foo')``.
+        :param prohibit_class_literals_before_attributes:
+            When ``True``, prefer all attribute blocks to be written before
+            any class literals.
+            For example: If set to ``True``, prefer
+            ``input(type='text').class`` over ``input.class(type='text')``.
```
Comment (Makman2) [inline bears/pug/PugLintBear.py]: what's with enforcing the other way around?

File: bears/pug/PugLintBear.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+                'maximumNumberOfLines': max_lines_per_file
+            }
+
+            return json.dumps(options)
+
+    @staticmethod
+    def create_arguments(filename, file, config_file, puglint_config: str=''):
+        """
+        :param puglint_config:
+            The location of the ``.pug-lintrc`` config file.
```
Comment (Makman2) [inline bears/pug/PugLintBear.py]: Maybe
```
The location of a custom  ``.pug-lintrc`` config file.
```
So it doesn't sound that this parameter is mandatory?

File: bears/pug/PugLintBear.py (hunk 9 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Done :+1:

File: bears/pug/PugLintBear.py (hunk 11 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Sounds good :+1:

File: tests/pug/puglint_test_files/.pug-lintrc [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+   "requireLowerCaseTags":true,
+   "disallowSpecificTags":null,
+   "requireClassLiteralsBeforeAttributes":false,
+   "disallowClassLiteralsBeforeAttributes":true,
+   "disallowIdLiteralsBeforeAttributes":true,
+   "disallowSpacesInsideAttributeBrackets":true,
+   "requireClassLiteralsBeforeIdLiterals":false,
+   "validateDivTags":true,
+   "disallowTagInterpolation":true
+}
```
Comment (yash-nisar) [inline tests/pug/puglint_test_files/.pug-lintrc]: Ah, eof damnit.

File: bears/pug/PugLintBear.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+                        prohibit_tag_interpolation: bool=True,
+                        prohibit_specific_attributes: typed_list(str)=None,
+                        prohibit_specific_tags: typed_list(str)=None,
+                        enforce_class_literals_before_attributes: bool=False,
+                        enforce_class_literals_before_id_literals: bool=False,
+                        enforce_id_literals_before_attributes: bool=False,
+                        enforce_lower_case_attributes: bool=True,
+                        enforce_lower_case_tags: bool=True,
+                        require_spaces_inside_attribute_brackets: bool=False,
+                        require_strict_equality_operators: bool=True,
```
Comment (Makman2) [inline bears/pug/PugLintBear.py]: what's with these using `enforce` as prefix?

**Comment by Asnelchristian** [comment]
consistency

2017-07-18 21:42 GMT+01:00 Mischa Krüger <notifications@github.com>:

**Comment by Asnelchristian** [comment]
we use more `enforce_` than `require_` :sweat_smile:

2017-07-18 22:03 GMT+01:00 Ngoulla Asnel Christian <notifications@github.com

File: bears/pug/PugLintBear.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+            ``input(type='text').class`` over ``input.class(type='text')``.
+        :param prohibit_class_literals_before_id_literals:
+            When ``True``, prefer all ID literals to be written before any
+            class literals.
+            For example: If set to ``True``, prefer
+            ``input#id.class(type='text')`` over
+            ``input.class#id(type='text')``.
+        :param prohibit_class_literals:
+            When ``True``, disallow any class literals.
+            For example: If set to ``True``, prefer ``div(class='class')``
```
Comment (Asnelchristian) [inline bears/pug/PugLintBear.py]: i am kind of lost ``prohibit_class_attribute_with_static_value`` default value is True so you make use of class literals but here ``prohibit_class_literals`` you prohibit them.... isn't that contradicting?

File: bears/pug/PugLintBear.py (hunk 13 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Okay, I'll use it for them too :+1:

File: bears/pug/PugLintBear.py (hunk 14 above)
Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: Fixed :+1:

**Comment by yash-nisar** [comment]
One weird thing that I noticed was that the config (which is later converted to json) only takes the values `true` and `null`, so I've made the required changes.

**Comment by yash-nisar** [comment]
CC @Makman2 @Asnelchristian

File: bears/pug/PugLintBear.py [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+        :param max_lines_per_file:
+            Number of lines allowed per file.
+        """
+        if puglint_config:
+            return None
+        else:
+            options = {
+                'disallowBlockExpansion':
+                    prohibit_block_expansion if prohibit_block_expansion
+                    else None,
```
Comment (Makman2) [inline bears/pug/PugLintBear.py]: can you maybe loop over the dict and assign `None` for each entry if needed?
```
options = {...}
for k, v in options.items():
    options[k] = v if v else None
```

Comment (Makman2) [inline bears/pug/PugLintBear.py]: or write a little function like `map_bool()`

Comment (yash-nisar) [inline bears/pug/PugLintBear.py]: I had initially done a `settings_map` where I mapped `False` to `None`

**Comment by yash-nisar** [comment]
@Makman2 Updated.

**Comment by Asnelchristian** [comment]
:+1:

**Comment by Asnelchristian** [comment]
ack https://github.com/coala/coala-bears/pull/1936/commits/4cfe9075758464864bbfe69b17d813ac78d171fb

**Comment by Asnelchristian** [comment]
@Makman2 ping! :smile:

**Comment by Makman2** [comment]
@Asnelchristian bam! :boom:
thanks for reviewing :D rebase and lgtm @yash-nisar :)

**Comment by yash-nisar** [comment]
reack 1d3182700dc80785ba7be703f439540c156134fc

**Comment by sils** [comment]
@rultor merge

**Pull request merged**